examples/fluids/src/setupdm.c

```diff
@@ -38,17 +39,43 @@ PetscErrorCode CreateDM(MPI_Comm comm, ProblemData *problem, MatType mat_type, V
 }
 
 // Setup DM
-PetscErrorCode SetUpDM(DM dm, ProblemData *problem, PetscInt degree, SimpleBC bc, Physics phys) {
+PetscErrorCode SetUpDM(DM dm, ProblemData *problem, PetscInt degree, PetscInt q_extra, SimpleBC bc, Physics phys) {
+  PetscInt q_order = degree + q_extra;
   PetscFunctionBeginUser;
   {
+    PetscBool is_simplex = PETSC_TRUE;
+
+    // Check if simplex or tensor-product mesh
+    PetscCall(DMPlexIsSimplex(dm, &is_simplex));
     // Configure the finite element space and boundary conditions
     PetscFE  fe;
     PetscInt num_comp_q = 5;
     DMLabel  label;
-    PetscCall(PetscFECreateLagrange(PETSC_COMM_SELF, problem->dim, num_comp_q, PETSC_FALSE, degree, PETSC_DECIDE, &fe));
+    PetscCall(PetscFECreateLagrange(PETSC_COMM_SELF, problem->dim, num_comp_q, is_simplex, degree, q_order, &fe));
     PetscCall(PetscObjectSetName((PetscObject)fe, "Q"));
     PetscCall(DMAddField(dm, NULL, (PetscObject)fe));
     PetscCall(DMCreateDS(dm));
+    {  // Project coordinates to enrich quadrature space
+      DM             dm_coord;
+      PetscDS        ds_coord;
+      PetscFE        fe_coord_current, fe_coord_new;
+      PetscDualSpace fe_coord_dual_space;
+      PetscInt       fe_coord_order, num_comp_coord;
+
+      PetscCall(DMGetCoordinateDM(dm, &dm_coord));
+      PetscCall(DMGetCoordinateDim(dm, &num_comp_coord));
+      PetscCall(DMGetRegionDS(dm_coord, NULL, NULL, &ds_coord, NULL));
+      PetscCall(PetscDSGetDiscretization(ds_coord, 0, (PetscObject *)&fe_coord_current));
+      PetscCall(PetscFEGetDualSpace(fe_coord_current, &fe_coord_dual_space));
+      PetscCall(PetscDualSpaceGetOrder(fe_coord_dual_space, &fe_coord_order));
+
+      // Create FE for coordinates
+      fe_coord_order = 1;
+      PetscCall(PetscFECreateLagrange(PETSC_COMM_SELF, problem->dim, num_comp_coord, is_simplex, fe_coord_order, q_order, &fe_coord_new));
```

Comment:
Why hard-code a lower order coordinates than the mesh might actually have?
Also, will you want to use faces in the future?
```suggestion
      PetscCall(PetscFECreateLagrange(PETSC_COMM_SELF, problem->dim, num_comp_coord, is_simplex, fe_coord_order, q_order, &fe_coord_new));
      PetscCall(PetscFEGetHeightSubspace(fe_coord_new, 1, &fe_coord_face_new));
```

Reply:
> Why hard-code a lower order coordinates than the mesh might actually have?

Because we haven't tested with higher-order geometry yet I guess? There are other things that will need to change before we get there. Probably better to warn/error out than hard code though.

> Also, will you want to use faces in the future?

Not sure what you mean by that, but we do use faces for boundary integrals. Also not sure what `PetscFEGetHeightSubspace` does without actually doing anything with `fe_coord_face_new`. Does that need an additional call to `DMProjectCoordinates(dm, fe_coord_face_new)`?

Reply:
That call sets up the FE on the faces here so it ready when you need it.